### DISCUSSION TITLE
Assert.AllAsync?

### DISCUSSION THREAD

**Discussion by safakgur** [xunit/xunit#2498] [Ideas]
We recently had a case where we wanted to do `Assert.All` and run async code for each item. SO has [this answer][so] that doesn't really work like `All` as it can fail on `Task.WhenAll`, so we don't get `All`'s nice per-item reporting.

Is this a scenario where there is already a recommended approach, or do you think adding an async version of `All` would be worth the investment?

Here's my simple version that runs all tests in sequence. I could split getting the tasks and awaiting them into separate loops, making the tests run in parallel, but I thought sequential would be a saner default.

```cs
public static async Task AllAsync<T>(IEnumerable<T> collection, Func<T, Task> action)
{
    GuardArgumentNotNull("collection", collection);
    GuardArgumentNotNull("action", action);

    var array = collection.ToArray();
    var errors = new Stack<Tuple<int, object?, Exception?>>();

    for (var i = 0; i < array.Length; i++)
        try
        {
            await action(array[i]);
        }
        catch (Exception ex)
        {
            errors.Push(new(i, array[i], ex));
        }

    if (errors.Count > 0)
        throw new AllException(array.Length, errors.ToArray());
}
```

[so]: https://stackoverflow.com/a/38597416/704144 "xUnit Assert.All() async"

**Comment by bradwilson** [comment]
This has already been implemented, and should be available in a v2.4.2 pre-release or v3 pre-release on MyGet. https://github.com/xunit/assert.xunit/blob/43e3e208a597bc1a4c77f2665f39f98e7401c157/CollectionAsserts.cs#L79-L137
